File: plugin/vscode/src/cli/Subprocess.ts
import { spawn } from "node:child_process";

export interface SubprocessResult {
  exit: number;
  out: string;
}

/** Run a command and capture merged stdout/stderr as utf-8 text. */
export function runSubprocess(
  command: string[],
  workingDirectory?: string,
): Promise<SubprocessResult> {
  return new Promise((resolve, reject) => {
    const [executable, ...args] = command;
    const proc = spawn(executable, args, {
      cwd: workingDirectory,
      env: { ...process.env, _DOCSIG_FORMAT_JSON: "true" },
      stdio: ["ignore", "pipe", "pipe"],
    });

    const chunks: Buffer[] = [];
    const onData = (chunk: Buffer) => chunks.push(chunk);

    proc.stdout.on("data", onData);
    proc.stderr.on("data", onData);

    proc.on("error", reject);
    proc.on("close", (code) => {
      resolve({
        exit: code ?? 1,
        out: Buffer.concat(chunks).toString("utf8").trim(),
      });
    });
  });
}

File: plugin/neovim/lua/docsig/subprocess.lua
local M = {}

function M.run_sync(command, cwd)
  local env =
    vim.tbl_extend("force", vim.fn.environ(), { _DOCSIG_FORMAT_JSON = "true" })

  -- stdout/stderr default to capture; `stdout = true` asserts on nvim 0.10
  local result = vim.system(command, { cwd = cwd, env = env }):wait()

  local out = (result.stdout or "") .. (result.stderr or "")
  return { exit = result.code, out = vim.trim(out) }
end

function M.run(command, cwd, on_complete)
  local env =
    vim.tbl_extend("force", vim.fn.environ(), { _DOCSIG_FORMAT_JSON = "true" })

  return vim.system(command, { cwd = cwd, env = env }, function(result)
    local out = (result.stdout or "") .. (result.stderr or "")
    on_complete({ exit = result.code, out = vim.trim(out) })
  end)
end

return M
